### ISSUE TITLE
IngressClass declared instead of 'kubernetes.io/ingress.class' annotation doesn't configure Ingress status

### ISSUE THREAD

**Issue by vugardzhamalov** [istio/istio#25308]
**Bug description**
I am trying to deploy 'Ingress' object with 'ingressClassName: istio' flag enabled instead of `kubernetes.io/ingress.class: istio` annotation as per [this](https://istio.io/latest/docs/tasks/traffic-management/ingress/kubernetes-ingress/#specifying-ingressclass) example in Istio traffic management documentation.

I am simply trying to reproduce this on my Minikube Cluster, but 'Ingress' object without annotation just won't work as expected. While it works with annotation - it doesn't work with 'ingressClassName: istio' flag.

Strangely enough if I happen to deploy Minikube's Ingress addon - it actually starts working -  but it doesn't go via Istion Ingress Gateway anymore - I am not sure if this is expected.  

[ ] Configuration Infrastructure
[X ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[X ] User Experience
[X ] Developer Infrastructure

**Expected behavior**
Running things on Minikube, Kubernetes version is 1.18.3 I was under impression that switching from `kubernetes.io/ingress.class: istio `annotation in favour of `ingressClassName: istio` flag won't result in any changes in terms of 'Ingress' objects not working without additional Ingress Controllers deployed. But it seems like it just won't pick-up Istio ingress controller at all... not sure if this is expected maybe I am missing the point.

**Steps to reproduce the bug**
1. Create Istio IngressClass object

  ```
apiVersion: networking.k8s.io/v1beta1
  kind: IngressClass
  metadata:
    name: istio
  spec:
    controller: istio.io/ingress-controller
```

2. Create Ingress object

  ```
apiVersion: networking.k8s.io/v1beta1
  kind: Ingress

  metadata:
    name: 'ingress-hnb'
    namespace: 'istio-ingress-lab'

  spec:
    ingressClassName: istio

    rules:
      - host: 'fancy.example.com'
        http:
          paths:
            - path: '/*'
              backend:
                serviceName: 'fancy-service'
                servicePort: 9999
```

3. Apply Ingress object configuration and describe it - one should see following (if minikube Ingress addon(aka kubernetes ingress controller)) has been deployed):

kubectl describe --namespace=istio-ingress-lab ingresses.networking.k8s.io ingress-hnb

```
Annotations:               Events:
  Type                     Reason  Age    From                      Message
  ----                     ------  ----   ----                      -------
  Normal                   CREATE  3m43s  nginx-ingress-controller  Ingress istio-ingress-lab/ingress-hnb
  Normal                   UPDATE  3m12s  nginx-ingress-controller  Ingress istio-ingress-lab/ingress-hnb

```

As you can see FROM is set to 'nginx-ingress-controller' I am not sure if this is correct...

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

bin/istioctl version --remote
```
client version: 1.6.4
control plane version: 1.6.4
data plane version: 1.6.4 (1 proxies)
```

kubectl version
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:52:00Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
`bin/istioctl manifest apply --set profile=default --set addonComponents.prometheus.enabled=false`

**Environment where bug was observed (cloud vendor, OS, etc)**
minikube version
```
minikube version: v1.11.0
commit: 57e2f55f47effe9ce396cea42a1e0eb4f611ebbd
```

**Comment by howardjohn** [comment]
@vugardzhamalov we have a test for this exact case: https://github.com/istio/istio/blob/ea4f294029d67433ebf0dc9ad49ee40eb42e2aa4/tests/integration/pilot/ingress_test.go#L112.

What I think is happening here is `nginx-ingress` is not respecting the ingressClass field, and taking over control of the Ingress object when it should not. Based onthis, I think this is a bug in nginx (or old version?) not Istio. Does that line up with your observations?

**Comment by vugardzhamalov** [comment]
Thank you @howardjohn !

OK, I thought I should deploy new Minikube profile and remove Kubernetes ingress-nginx controller (Minikube addon) out of equation. My actions/observations are:

1. Deployed new Minikube profile (same version of Minikube and same version of Kubernetes as above) without Ingress addon enabled!

2. Deployed Istio 1.6.7 with istioctl exactly as per my earlier post.

3. Deployed 'Ingress' object with single annotation: 'kubernetes.io/ingress.class: istio' - it seems as if things working as expected.

4. Removed earlier deployed 'Ingress' object. Deployed a new one without annotations at all - but with 'ingressClassName: istio' flag declared instead... It seems like Ingress works - I can definitely reach out to my endpoints behind the Ingress!  But I wouldn't say that it is working as expected, or maybe I just don't understand the difference.... So my current observations are:

IMHO the most important difference observed when I issue:

`kubectl describe ingresses.networking.k8s.io --namespace=<your-ingress-namespace> <your-ingress>`

1. When you are doing this against Ingress object deployed with annotations normally you'll be given 'Address' that will correspond to Istio Ingress Gateway IP address. On top of that the host name(s) declared in your Ingress rule(s) will be created for you if you've set your 'external-dns' accordingly and this will end up reflected in 'external-dns' logs!

2. Now if you are issuing 'kubectl describe' against Ingress object deployed with 'ingressClassName: istio' flag instead of annotations - things starting to look somewhat different. First of all I don't see 'Address' filed populated at all... it just stays empty! The next thing is even more mysterious ... 'external-dns' while does! create required DNS resource records as per host name(s) in your Ingress rule(s) it does so quietly... like there is nothing in 'external-dns' logs at all! 

3. So I guess I could say it works for me! But I see no address assigned to it and for some reason (maybe related to missing IP address) 'external-dns' logging doesn't give any details on resource records being created (once again - these are quietly created). But maybe this is expected change in behaviour - I don't know. Also please keep in mind that I wasn't making any changes to verbosity in logging for my 'external-dns' instance.

Please let me know if you want me to make further changes/observations - happy to assist as much as I can. But again how many people will be even looking into this functionality with Ingress objects? Even documentation clearly states that we are better off with Istio Gateway and VirtualService objects instead. So if you think that this is good enough for now - I'll be happy to close the ticket.

**Comment by howardjohn** [comment]
@vugardzhamalov I see, thanks for the info! Looks like we are respecting the new `ingressClassName` field for the core Ingress code, which does the real logic of configuring routing, but the status code which is separate is not updated (https://github.com/istio/istio/blob/623cbcd5d7be9e0aa1e8aae8c66b74105dd9d848/pilot/pkg/config/kube/ingress/status.go#L291). So we will need to update this.

**Comment by bianpengyuan** [comment]
@howardjohn @ramaraochavali Hi this is a P0 for 1.8, please help to triage this issue thanks!

**Comment by howardjohn** [comment]
Dropping to P1

**Comment by howardjohn** [comment]
I wrote a failing test that covers this: https://github.com/istio/istio/compare/master...howardjohn:test/ingress-status-integ?expand=1

**Comment by Monkeyanator** [comment]
Fixed in #28225
